### PR TITLE
Implement custom search functions

### PR DESCRIPTION
**Pull request by lukasmalkmus**
I think there are many ways to implement the configuration proposal I made in #19. I made up this draft for the following reason:

The jwtauth package has two main components used by package users: The `Verifier` and the `Authenticator` functions. Implementing a custom `Authenticator` is easy because the provided one is ~10 lines of code. But implementing a custom `Verifier` is a little bit trickier because this is where the token gets extracted from the request, validated, etc. 

So instead of adding more function parameters or a custom options struct, I came up with this PR which adds so called `Extractors`. What sounds complicated is pretty easy: Instead of hard coding where the `Verifier` searches for the token string in the request, we provide it with `ExtractorFunc`s that do this task for the `Verifier`. Furthermore, the signature of the original `Verifier` function has not been changed, so there won't be any breaking changes for users relaying  on the default behaviour of this package.

What got removed are the `params`. But I don't think that's critical. They can be replaced by implementing a custom `ExtractorFunc`.

Furthermore there was no need to adjust the tests.

So what got added in this PR?
* `ExtractorFunc` type
* Three extractors (functions): JWT Cookie, Authroization Header, JWT query param

I think it's best if @pkieltyka or another maintainer takes a look at the code and provides me with some feedback.

### CODE REVIEW

**Comment by lukasmalkmus** [comment]
I think it is also worth mentioning that this PR preserves the token search order. If a users implements the default `Verifier` middleware the request is searched for the token in the URL first, then in the "Authorization" header and finally in the "jwt" cookie.

But this PR also fixes #5 by implementing the `Verify` function and passing extractors in the desired order: 
```go
jwtauth.Verify(auth, jwtauth.ExtractCookie, jwtauth.ExtractHeader, jwtauth.ExtractQuery)
```

**Comment by pkieltyka** [comment]
well done. I like this a lot - only feedback is can we think of a better verb than "Extractor" / "Extract" ? cc: @VojtechVitek 

I'll think as well

**Comment by lukasmalkmus** [comment]
Right, `Extractor` and `ExtractorFunc` don't fit very good. Because the function signature of the `ExtractorFunc` type is so short, we could just omit it. `func(*http.Request) string` isn't that complicated.
Maybe a good name would be `From...` for the "search" functions:
`jwtauth.Verify(auth, jwtauth.FromCookie, jwtauth.FromHeader, jwtauth.FromQuery)`

File: jwtauth.go [hunk 1]
```diff
@@ -21,6 +21,36 @@ var (
 	ErrExpired      = errors.New("jwtauth: token is expired")
 )
 
+// An ExtractorFunc retreives the token string from the request.
+type ExtractorFunc func(r *http.Request) string
```
Comment (lukasmalkmus) [inline jwtauth.go]: We could omit the new type definition here

File: jwtauth.go [hunk 2]
```diff
@@ -68,53 +98,33 @@ func NewWithParser(alg string, parser *jwt.Parser, signKey []byte, verifyKey []b
 // http response.
 func Verifier(ja *JwtAuth) func(http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
-		return Verify(ja, "")(next)
+		return Verify(ja, ExtractQuery, ExtractHeader, ExtractCookie)(next)
 	}
 }
 
-func Verify(ja *JwtAuth, paramAliases ...string) func(http.Handler) http.Handler {
+func Verify(ja *JwtAuth, extractors ...ExtractorFunc) func(http.Handler) http.Handler {
```
Comment (lukasmalkmus) [inline jwtauth.go]: And just use `func(r *http.Request) string` as parameter, which isn't much larger or complicated:
`func Verify(ja *JwtAuth, tokenSources ...func(r *http.Request) string) func(http.Handler) http.Handler {...}`

File: jwtauth.go [hunk 3]
```diff
@@ -21,6 +21,36 @@ var (
 	ErrExpired      = errors.New("jwtauth: token is expired")
 )
 
+// An ExtractorFunc retreives the token string from the request.
+type ExtractorFunc func(r *http.Request) string
+
+var (
+	// ExtractCookie tries to retreive the token string from a cookie named "jwt".
+	ExtractCookie = func(r *http.Request) string {
```
Comment (pkieltyka) [inline jwtauth.go]: How about names like..

`FindCookieToken = func(r *http.Request) string`, `FindHeaderToken` etc..

and as you suggested, we can type the func type and in the constructor just have `findTokens ...func(*http.Request) string`

Comment (pkieltyka) [reply jwtauth.go]: or `findTokenFns ...func(*http.Request) string`

**Comment by pkieltyka** [comment]
thanks @VojtechVitek for the feedback. I think `FromCookie` isn't clear, it should be clear we're looking for a token from a http.Request, within the cookie. `FindCookieToken` etc. is more clear

**Comment by lukasmalkmus** [comment]
I implemented the discussed changes. Is the naming ok? I decided to go with `TokenFrom...`. Feedback on this pattern? I only clashes with the `TokenCtxKey` but that should be ok.

**Comment by pkieltyka** [comment]
hey @lukasmalkmus my apologies for the delay - I was away on vacation for a few weeks. I've reviewed it again and looks good, going to merge! thank you again for this awesome contribution.

**Comment by lukasmalkmus** [comment]
Nothing to apologise for :) Thanks for merging. I just pulled in the latest release and as expected, nothing to change in my code, since the package API did not change 🎉 

Also closed #19

**Pull request merged**